### PR TITLE
Improve template and error message and  documentation

### PR DESCRIPTION
**Pull request by yoavkatz**

### CODE REVIEW

File: docs/docs/adding_template.rst [hunk 1]
```diff
@@ -56,9 +56,43 @@ We can define a template for this task like this:
         output_format='{translation}',
     ),
 
-Important: the only fields that are mandatory are the `input_format` and `output_format`; without them, unitxt won't know how to use the task fields.
 
-Templates for Special Data
+The `instruction` attribute defines that part of the prompt that appears one, while the `input_format` defines the part of prompt that repeats for
```
Comment (yoavkatz) [inline docs/docs/adding_template.rst:60]:
```suggestion
The `instruction` attribute defines that part of the prompt that appears once, while the `input_format` defines the part of prompt that repeats for 
```

File: docs/docs/adding_template.rst [hunk 2]
```diff
@@ -56,9 +56,43 @@ We can define a template for this task like this:
...
     ),
 
-Important: the only fields that are mandatory are the `input_format` and `output_format`; without them, unitxt won't know how to use the task fields.
 
-Templates for Special Data
+The `instruction` attribute defines that part of the prompt that appears once, while the `input_format` defines the part of prompt that repeats for 
+each in-context learning demonstration and for the final instance.  The `output_format` defines how the reference answer is verbalized as string.   
+The InputOutputTemplate assumes there is at most only a single reference (gold answer). 
+If you pass a field value which is a list to the InputOutputTemplate, then it is verbalized as comma separated string. For example, ["happy","angry"]
+become the string reference "happy,angry", and it is expected that the model will return that string as the correct answer.
```
Comment (yoavkatz) [inline docs/docs/adding_template.rst:64]:
```suggestion
becomes the string reference "happy,angry", and it is expected that the model will return that string as the correct answer.
```

File: docs/docs/adding_template.rst [hunk 3]
```diff
@@ -56,9 +56,43 @@ We can define a template for this task like this:
...
+.. note::
+    If you don't have references , just specify `output_format=""`.
+    If you have multiple references, use the MultiReferenceTemplate (see below)
+    The only fields that are mandatory are the `input_format` and `output_format`
+
+Post Processors
+---------------
+
+The template also defines the post processing steps done on the output predictions of the model before they are passed to the :ref:`Metrics <metric>`.
+The post processors applied both to the textual prediction and to the references.
```
Comment (yoavkatz) [inline docs/docs/adding_template.rst:75]: [](url)
```suggestion
The post processors applied both to the model prediction and to the references. 
```

File: docs/docs/adding_template.rst [hunk 4]
```diff
@@ -56,9 +56,43 @@ We can define a template for this task like this:
...
+    If you have multiple references, use the MultiReferenceTemplate (see below)
+    The only fields that are mandatory are the `input_format` and `output_format`
+
+Post Processors
+---------------
+
+The template also defines the post processing steps done on the output predictions of the model before they are passed to the :ref:`Metrics <metric>`.
+The post processors applied both to the model prediction and to the references. 
+For example, we could use the `processors.lower_case` processor to lowercase both the model predictions and references,
+so the metrics computation will ignore case.
```
Comment (yoavkatz) [inline docs/docs/adding_template.rst:77]:
```suggestion
so the metric computation will ignore case. 
```

File: docs/docs/adding_template.rst [hunk 5]
```diff
@@ -56,9 +56,43 @@ We can define a template for this task like this:
...
+
+    template = InputOutputTemplate(
+        instruction="In the following task, you translate a {text_type}.",
+        input_format="Translate this {text_type} from {source_language} to {target_language}: {text}.",
+        target_prefix="Translation: ",
+        output_format='{translation}',
+        postprocessors= [
+                "processors.lower_case"
+            ]
+    )
```
Comment (elronbandel) [inline docs/docs/adding_template.rst:87]: should be in code block

File: docs/docs/adding_template.rst [hunk 6]
```diff
@@ -56,9 +56,43 @@ We can define a template for this task like this:
...
     ),
 
-Important: the only fields that are mandatory are the `input_format` and `output_format`; without them, unitxt won't know how to use the task fields.
 
-Templates for Special Data
+The `instruction` attribute defines that part of the prompt that appears once, while the `input_format` defines the part of prompt that repeats for 
+each in-context learning demonstration and for the final instance.  The `output_format` defines how the reference answer is verbalized as string.   
+The InputOutputTemplate assumes there is at most only a single reference (gold answer). 
+If you pass a field value which is a list to the InputOutputTemplate, then it is verbalized as comma separated string. For example, ["happy","angry"]
+becomes the string reference "happy,angry", and it is expected that the model will return that string as the correct answer.
```
Comment (elronbandel) [inline docs/docs/adding_template.rst:64]:
```suggestion
The ``instruction`` attribute defines that part of the prompt that appears once, while the ``input_format`` defines the part of prompt that repeats for 
each in-context learning demonstration and for the final instance.  The ``output_format`` defines how the reference answer is verbalized as string.   
The InputOutputTemplate assumes there is at most only a single reference (gold answer). 
If you pass a field value which is a list to the InputOutputTemplate, then it is verbalized as comma separated string. For example, ["happy","angry"]
becomes the string reference "happy,angry", and it is expected that the model will return that string as the correct answer.
```

File: docs/docs/adding_template.rst [hunk 7]
```diff
@@ -56,9 +56,43 @@ We can define a template for this task like this:
...
+    If you have multiple references, use the MultiReferenceTemplate (see below)
+    The only fields that are mandatory are the `input_format` and `output_format`
+
+Post Processors
+---------------
+
+The template also defines the post processing steps done on the output predictions of the model before they are passed to the :ref:`Metrics <metric>`.
+The post processors applied both to the model prediction and to the references. 
+For example, we could use the `processors.lower_case` processor to lowercase both the model predictions and references,
+so the metric computation will ignore case.
```
Comment (elronbandel) [inline docs/docs/adding_template.rst:77]:
```suggestion
The template also defines the post processing steps applied to the output predictions of the model before they are passed to the :ref:`Metrics <metric>`.
The post processors applied both to the model prediction and to the references. 
For example, we could use the ``processors.lower_case`` processor to lowercase both the model predictions and references,
so the metric computation will ignore case. When needed, It is possible to add post processors that applied only to the output of the model and not the references or vice versa. 
```

File: docs/docs/adding_template.rst [hunk 8]
```diff
@@ -56,9 +56,43 @@ We can define a template for this task like this:
...
+        input_format="Translate this {text_type} from {source_language} to {target_language}: {text}.",
+        target_prefix="Translation: ",
+        output_format='{translation}',
+        postprocessors= [
+                "processors.lower_case"
+            ]
+    )
+
+The reason the post processors are set in the template, is because different templates prompt the model to generate answers in different formats. 
+For example, one template may prompt the model to answer `Yes`` or `No` while another
```
Comment (elronbandel) [inline docs/docs/adding_template.rst:90]:
```suggestion
For example, one template may prompt the model to answer `Yes` or `No` while another 
```

**Review by elronbandel — APPROVED** [review]
There are few minor fixes, other than that it's good to go.

File: docs/docs/adding_template.rst (hunk 5 above)
Comment (yoavkatz) [reply docs/docs/adding_template.rst:96]: Done.

File: docs/docs/adding_template.rst [hunk 9]
```diff
@@ -56,9 +60,48 @@ We can define a template for this task like this:
...
+        target_prefix="Translation: ",
+        output_format='{translation}',
+        postprocessors= [
+                "processors.lower_case"
+            ]
+    )
+
+The reason the post processors are set in the template, is because different templates prompt the model to generate answers in different formats. 
+For example, one template may prompt the model to answer ``Yes`` or ``No`` while another 
+template may prompt the model to answer ``True`` or ``False``. Both can use different post processors to convert them to standard model prediction of `0` or `1`
```
Comment (yoavkatz) [inline docs/docs/adding_template.rst:100]:
```suggestion
template may prompt the model to answer ``True`` or ``False``. Both can use different post processors to convert them to standard model prediction of `0` or `1`.
```

**Pull request merged**